---
title: "Welcome & Introduction"
subtitle: "ASI: Introduction to R"
author: "Dr Stevie Pederson"
institute: |
  | Black Ochre Data Labs
  | Telethon Kids Institute
date: "2025-09-2"
date-format: long
title-slide-attributes:
    data-background-color: "#3c3c44"
    data-background-image: assets/bodl_logo_white_background.jpg
    data-background-opacity: "0.3"
    data-background-size: "90%"
editor: source
format: 
  revealjs:
    theme: [bodl.scss]
    code-line-numbers: false
    code-tools: true
    logo: assets/bodl_logo_white_background.jpg
    height: 720
    width: 1280
    sansfont: Times New Roman
    slide-number: c
    show-slide-number: all
  html: 
    css: [bodl.scss, extra.css]
    output-file: intro.html
    embed-resources: true    
    toc: true
    toc-depth: 2
include-after: |
  <script type="text/javascript">
    Reveal.on('ready', event => {
      if (event.indexh === 0) {
        document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
      }
    });
    Reveal.addEventListener('slidechanged', (event) => {
      if (event.indexh === 0) {
        Reveal.configure({ slideNumber: null });
        document.querySelector("div.has-logo > img.slide-logo").style.display = "none";
      }
      if (event.indexh === 1) { 
        Reveal.configure({ slideNumber: 'c' });
        document.querySelector("div.has-logo > img.slide-logo").style.display = null;
      }
    });
  </script>    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, include = TRUE, warning = FALSE, message = FALSE, 
  fig.align = "center"
)
```


<br><br><br>

### http://blackochrelabs.au/20250902_ASI_RWorkshop/ 


# Introduction {background-color="#3c3c44"}

---

We would like to acknowledge that today we're meeting on the lands of the Noongar-Whadjuk people. 
We acknowledge the deep feelings of attachment and relationship of the Noongar people to their Place.

We also pay respects to the cultural authority of Aboriginal and Torres Strait Islander peoples from other areas of Australia today, and pay our respects to Elders past, present and emerging. 

::: {.fragment}

As a scientist, I'm honoured to be working with many Indigenous Australians, who have a proud history of being some of the first acknowledged scientists from the emerging historical record 

:::

## Who Am I? 

**Stephen (Stevie) Pederson** (They/Them)

- Adelaide, Kaurna Country, SA
- Bioinformatician, Black Ochre Data Labs, The Kids Research Institute Australia

::: {.fragment}

- Bioinformatician, Dame Roma Mitchell Cancer Research Laboratories (2020-2022)
- Co-ordinator, UofA Bioinformatics Hub (2014-2020)

:::

<!-- . . . -->

<!-- - PhD (2008-2018) was a Bayesian Model for Transcript-Level Analysis -->
<!--     + MCMC Engine written in C & R (No Rstudio. No `Rcpp`. No `RStan`) -->
<!--     + Microrray analysis (T~reg~ & T~h~ Activation) -->
    
::: {.notes}
Used to write Excel macros for fun whilst being a musician
:::

## Who Am I? {.slide-only .unlisted}

**Stephen (Stevie) Pederson** (They/Them)

- R User for ~20 years $\implies$ learnt when R was difficult!
- Senior Author of 7 Bioconductor Packages 
    + `ngsReports`, `extraChIPs`, `transmogR`, `motifTestR`
    + `strandCheckR`, `sSNAPPY`, `tadar`
- Co-Chair, Bioconductor Community Advisory Board

::: {.fragment}

*Made countless typos, horrible decisions and catastrophic errors*

:::

## Today's Tutors

- Dr Sam Buckberry (Black Ochre Data Labs)
- Dr Jennifer Currenti, Eric Alves (Harry Perkins Institute for Medical Research)


## Housekeeping

- Toilets
- Morning Tea

    
## Homepage and Material

- The workshop homepage is [http://blackochrelabs.au/20250902_ASI_RWorkshop/](http://blackochrelabs.au/20250902_ASI_RWorkshop/)
    - Data and course material available here
    + Will stay live *in perpetuity*
- Links to notes available
    + Slides are directly re-formatted as a simple webpage
    + Also in presentation style by clicking the `RevealJS` link below the TOC<br>(Top RH Corner)


# A Brief Introduction to R {background-color="#3c3c44" background-image=https://upload.wikimedia.org/wikipedia/commons/thumb/1/1b/R_logo.svg/512px-R_logo.svg.png background-size="40%" background-opacity="0.5" background-position='80% 50%'}


## Why use R?

- Heavily used for analysis of biological data (along with Python)
    - Can handle extremely large datasets  
    - Packages explicitly designed for complex analysis
    - Huge user base of biological researchers (www.bioconductor.org)
    - (Can be) very fast
- Very easy to *dynamically interact* with large datasets
    - Can also run as scripts on HPC clusters
    
::: {.notes}
I regularly work with data containing millions of lines
:::

## Why use R? {.slide-only .unlisted}
    
- Never use drop-down menus or buttons $\implies$ type every command
    - Complex processes are often implemented as 'simple' functions

::: {.fragment}
    
- **Reproducible Research!!!**
    - Transparent methods
    - Integration with version control such as `git`
- Avoids common Excel pitfalls [(almost) never modify files on disk!]{.fragment}

:::

::: {.fragment}

Experience is the best teacher $\implies$ please practice your skills

:::

::: {.notes}
- Discuss column sorting
:::

    
## What is R?

- Derivative of `S` (*John Chambers et al*, Bell Labs 1976)
- `R` first appeared in 1993
    - **R**oss Ihaka and **R**obert Gentleman (U of Auckland)
    - Authors wrote for their own research and students
    - First official release (v1.0.0) in Feb 2000
    - Now estimated >2 million users (2012)

::: {.notes}
- Ross Ihaka is of NZ Maori descent
:::    
    
## What is R? {.slide-only .unlisted}

::: {.columns}

:::: {.column width='75%'}

- Open source language
    + No corporate ownership $\implies$ free software
    + Code is managed by the community of users
- `R` is formally run by a volunteer committee (**R Core**)
    + Mostly academics
    + John Chambers is *still* a member
- Annual release schedule with patches as required

::::

:::: {.column width='25%'}

![Extending R, Chambers (2016)](https://images.routledge.com/common/jackets/amazon/978149877/9781498775717.jpg){width="250px"}

::::

:::

::: {.notes}
- Being open source creates headaches for University & Business IT departments
- No guarantees of being virus free
- The community self-regulates
:::
    
## R Packages

- Packages are the key to R's flexibility and power
    - Are collections (or libraries) of related functions
    - `ggplot2` $\implies$ Create plots using the *grammar of graphics*
    - `readr` $\implies$ Read files into `R
    
## R Packages {.slide-only .unlisted}
    
::: {.notes}
- Loading a package is like the Matrix where Trinity installs the set of skills to fly a helicopter
- Crap packages are generally identified by the users and just not-used
- Scientific rigour is usually checked during review of the accompanying publication
- Robert Gentleman was also a founder of Bioconductor before moving to 23&me
:::    
    
- $>$ 22,000 packages are stored on CRAN (https://cran.r-project.org)
    - Not curated for statistical quality or documentation
    - Automated testing for successful installation

::: {.fragment}

- Bioconductor is a secondary repository (https://www.bioconductor.org)
    - $>$ 3000 packages with a more biological/genomics focus
    - Curated for language consistency & documentation

:::

::: {.fragment}
- Some packages also live on github or bitbucket
    + Will include latest (unstable) development versions
    
:::


## R Packages {.slide-only .unlisted}

- Installation of a package can be done through a variety of methods:
    + `install.packages("packageName")` will only install from CRAN

::: {.fragment}
- Bioconductor also provide a CRAN package: `BiocManager`
    + `BiocManager::install("packageName")` installs from CRAN, Bioconductor & github

:::

::: {.fragment}
- The CRAN package `pak` also installs from multiple locations
:::


## Helpful Resources {visibility="hidden"}



::: {.notes}
Much of today is inspired by a two-day developers workshop I attended with Hadley Wickham.
Also gave me an opportunity to have some great conversations with Winston Chang
:::

<br>

::: {.columns}

::: {.column width="50%}


![https://r4ds.had.co.nz/](https://m.media-amazon.com/images/I/51eTKL+SkeL.jpg){width="350px" height="495px" fig-align="left"}

:::

::: {.column width="50%}


![https://r-graphics.org/](https://r-graphics.org/cover.jpg){width="350px" height="495px" fig-align="left"}


:::

:::



# Using R {background-color="#3c3c44" background-image=https://upload.wikimedia.org/wikipedia/commons/thumb/1/1b/R_logo.svg/512px-R_logo.svg.png background-size="40%" background-opacity="0.5" background-position='80% 50%'}

## The R Console

::: {.notes}
- This is the 'ugly, old-school' way of using R
- Still very useful, e.g. testing code on a HPC
:::


:::: {.columns}
::: {.column width="55%"}

- Let's try using `R` as a standalone tool $\implies$ open `R` NOT `RStudio`
    + On `linux`: <br>Open a terminal then enter `R`
    + OSX: Click ![](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1b/R_logo.svg/512px-R_logo.svg.png){width="40px"} on your dock
    + Windows: Click ![](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1b/R_logo.svg/512px-R_logo.svg.png){width="40px"} in your Start Menu
    
    
- Do not open ![](assets/RStudio_icon.png){width="60px"} 
- This is `R` at it's ugliest (how I learned R)

:::
::: {.column width="45%"}

![](assets/RTerminal.png){.absolute right=0 width="550px"}

:::

::::

## The R Console {.slide-only .unlisted}

- This is often referred to as the `R Console` 
- At it's simplest is just a calculator

```{r}
1 + 1
2 * 2
2 ^ 3
```

## The R Console {.slide-only .unlisted}


- `R` has many standard functions

```{r}
sqrt(2)
log10(1000)
```

::: {.fragment}

- NB: We placed the value *inside* the brackets *after* the function name

:::

::: {.notes}
I never use a calculator program on my laptop, always R
:::

## The R Console {.slide-only .unlisted}

We can create objects with names

```{r}
x <- 5
```

::: {.fragment}

- We have just created an object called `x`
- View the contents of the object by entering it's name in the `Console`

```{r}
x
```

:::

::: {.fragment}

- The `<-` symbol is like an arrow $\implies$ "put the value `5` into `x`"

:::
::: {.fragment}

![*An APL Keyboard from the 1970s*](assets/APL-keybd2.png){width=60% fig-align="left"}
:::

## The R Console {.slide-only .unlisted}

- The object `x` only exists in the `R Environment`

::: {.fragment}

- We can pass objects to functions and perform operations on the

```{r}
x + 1
sqrt(x)
x^2
x > 1
```

:::

## The R Console {.slide-only .unlisted}

- Everything we've just done is trivial
- Real analysis isn't
- If we perform a series of steps
    + Should we keep a copy of what we've done?
    + If so, how should we do that?
    
::: {.fragment}

- A common strategy is to record our code as an *R Script*
- `R Studio` makes that easy & convenient

:::
